MAINT: Implement new style promotion for `np.result_type`, etc.

seberg opened the pull request:
This implements a new style promotion scheme for mutiple inputs
as are passed to `np.result_type`. It should be noted that
result-type heavily relies on "value-based casting" logic/promotion
logic.
This fact is inherited to many places.  When user DTypes are introduced
some of these places should potentially use "weak DType" logic instead
if the input is a Python scalar.
This is *not* yet implemented. It would be necessary to mark these
the same way as is currently done for `np.result_type`.

This type of "marking" will generally be necessary in a few places,
even if may get away with just switching the logic everywhere.

---

Marking as draft, since I am planning to probably write some small tests (there are some exceedingly rare changes in behaviour).  Also, probably moving some of the new code to the new `common_dtype.c` file.  But I don't expect any larger changes right now, so putting it out there.

I expect the doc-tests will fail, this is OK, the doctests need to be adapted.  There is a _very_ mild slowdown in ufunc loop lookup due to this.  That seems OK, it probably is around the same or less than the general speedup in gh-15271.  Eventually it would largely go away again in the ufunc refactor (although the exact speed impact of that is unknown, we are talking about overheads of at most 10-20% of the cheapest possible ufunc calls...).

seberg committed:
MAINT: Implement new style promotion for `np.result_type`, etc.

This implements a new style promotion scheme for mutiple inputs
as are passed to `np.result_type`. It should be noted that
result-type heavily relies on "value-based casting" logic/promotion
logic.
This fact is inherited to many places.  When user DTypes are introduced
some of these places should potentially use "weak DType" logic instead
if the input is a Python scalar.
This is *not* yet implemented. It would be necessary to mark these
the same way as is currently done for `np.result_type`.

This type of "marking" will generally be necessary in a few places,
even if may get away with just switching the logic everywhere.

tylerjereddy commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -867,44 +867,44 @@ array_astype(PyArrayObject *self,
+    if ((PyArray_NDIM(self) != PyArray_NDIM(ret)) &&
+            DEPRECATE_FUTUREWARNING(
+                "casting an array to a subarray dtype "
+                "will not using broadcasting in the future, but cast each "
```
using -> use?

tylerjereddy commented on numpy/core/src/multiarray/abstractdtypes.c:
```diff
@@ -126,43 +140,147 @@ initialize_and_map_pytypes_to_dtypes()
+/*
+ * The following functions define the "common DType" for the abstract dtypes.
+ *
+ * Note that the logic with resupt to the "higher" dtypes such as floats
```
`resupt` -> `respect`?

tylerjereddy commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -856,6 +856,17 @@ typedef int (PyArray_FinalizeFunc)(PyArrayObject *, PyObject *);
+     * python float, int, or complex.
+     * An array using this flag must be a temporary array that can never
+     * leave the C internals of NumPy.  Even if it does, ENSUREARRAY is
+     * absolutely safe to abuse, since it alraedy is be a base class array :).
```
`alraedy` -> `already`?

`is be a base class` -> `is a base class`?

tylerjereddy reviewed (commented):
Was just reading through out of curiosity and added a few minor comments for typos and stuff.

seberg committed:
DOC: Fix mistakes found by Tyler

seberg commented:
Thanks Tyler, let me know if you have any comments. I should round this off (mainly add tests, and scan through this). Although, I will read through the NEP 43 draft first now, since promotion in ufuncs is also confusing me again and that document also needs to settle soon.

seberg committed:
DOC,TST: Fix documentation and add tests

seberg committed:
MAINT: Remove `NPY_USE_NEW_CASTINGIMPL`

This doesn't serve any purpose anymore. The new code is now always
used.  (In very few cases this may lead to small slowdowns, this
should only matter in ufuncs where it doesn't seem to matter enough
to worry about it.)

seberg committed:
TST: Fixup tests and make sure complex is also the value

seberg commented:
Ping @mattip if you have some time please have a look at this.

mattip reviewed (commented):
There are a few bits with complex dtypes that coverage says are not hit. Otherwise it all looks good to me.

seberg committed:
BUG: Fortify against longdouble cast loss

Base the "magic" value of the double, and use nextafter even for the
equal condition to avoid rounding during the cast to longdouble.
This might further be necessary because the code uses double constants
and not longdouble ones (not sure about this).

Using nextafter on the double value should put us firmly on the right
side and nobody should care about the few ulps in between those values.

seberg committed:
Apply suggestions from code review

Co-authored-by: Matti Picus <matti.picus@gmail.com>

seberg commented on numpy/core/src/multiarray/array_coercion.h:
```diff
@@ -15,6 +15,8 @@ typedef struct coercion_cache_obj {
 } coercion_cache_obj;
 
+NPY_NO_EXPORT PyArray_DTypeMeta *
+npy_discover_dtype_from_pytype(PyTypeObject *pytype);
```
Thanks, must have been some transitionary thing.

mattip replied:
Can we delete these lines if they are not needed?

seberg commented on numpy/core/src/multiarray/abstractdtypes.c:
```diff
@@ -13,6 +13,12 @@
+{
+    return PyArray_DescrFromType(NPY_LONG);
+}
+
```
All DTypes should have a `default_descr()` class method (even these abstract ones). I guess it could make sense to not have one in some cases, but here it is required.

seberg committed:
TST: Add complex-complex promotions to test parametrization

seberg commented:
Sorry for missing to update this more earlier.  I added some new tests parametrizations, hopefully that will clear out the missed coverage.

mattip commented on numpy/core/src/multiarray/abstractdtypes.c:
```diff
@@ -126,43 +140,147 @@ initialize_and_map_pytypes_to_dtypes()
+    if (other->legacy && other->type_num < NPY_NTYPES) {
+        if (other->type_num == NPY_BOOL) {
+            /* Use the default integer for bools: */
+            return PyArray_DTypeFromTypeNum(NPY_LONG);
```
Is there a NPY_DEFAULT_INT type? Also - is coverage correct that this case is not tested?

seberg replied:
Added the coverage.  No, we don't have such a macro.  There are two places that three places currently where we use a default I think:

* Here
* In this file, but in the `discover_descriptor_from_pyint`
* In the ufunc code when specialising `np.add.reduce` and `np.multiply.reduce` to always return integers.

mattip commented on numpy/core/src/multiarray/abstractdtypes.c:
```diff
@@ -126,43 +140,147 @@ initialize_and_map_pytypes_to_dtypes()
+    else if (other->legacy) {
+        /* This is a back-compat fallback to usually do the right thing... */
+        return PyArray_DTypeFromTypeNum(NPY_UINT8);
+    }
```
test needed?

mattip commented on numpy/core/src/multiarray/common_dtype.c:
```diff
@@ -0,0 +1,318 @@
+    if (length == 1) {
+        Py_INCREF(dtypes_in[0]);
+        return dtypes_in[0];
+    }
```
test missing?

seberg replied:
I think this one is impossible to reach from Python currently.  A bit of a bummer, but I would like to be pragmatic about it.

Finally, got the rest of the coverage right though.  (I was missing multiple Python scalars as inputs paths, probably that also were not also _all_ scalars).

mattip reviewed (commented):
I think this makes sense. There are many new codecov messages about code not hit, are they correct?

seberg committed:
TST: Add test for rational paths

(also delete those lines I forgot to delete)

seberg committed:
TST: Add more tests to cover abstract code paths (hopefully)

charris commented:
Thanks Sebastian.
